### PR TITLE
[Kimi-K3] Enable REAP expert pruning with LinearExperts2D

### LINKED ISSUE #2979: [Kimi-K3] Integrate KimiSparseMoeBlock with LinearExperts2D for REAP Support
## Background ##
The Kimi-K3 model definition currently uses `KimiSparseMoeBlock`, which does not integrate with LLM Compressor's `LinearExperts2D` class. This means the REAP algorithm cannot be applied to Kimi-K3 models, which is a significant missed opportunity for memory savings through expert pruning.

If `KimiSparseMoeBlock` is replaced with `LinearExperts2D`, the full REAP expert-removal flow becomes available for Kimi-K3 models.

## Proposed Changes ##
1. Follow the `inference-optimization/Kimi-K3-0.40B` model card to demonstrate loading the tiny version of the model using the `KimiK3ForConditionalGeneration` definition provided by LLM Compressor
2. Replace `KimiSparseMoeBlock` with `LinearExperts2D` in the model definition
3. Add a load mapping that maps weight names from the original `KimiSparseMoeBlock` class to the new `LinearExperts2D` class
4. Verify that the model can load with the new `LinearExperts2D` class and add a test to verify that outputs are the same before and after the replacement
5. Run the REAP algorithm on the tiny model to verify the end-to-end flow

### PR DESCRIPTION
**Pull request by KKothuri**
Closes #2979

## Summary

This PR integrates the Kimi-K3 routed experts with the `LinearExperts2D` representation required by REAP. It preserves the existing Kimi routing, shared-expert, latent-projection, and SiTU/SiLU behavior while exposing each routed expert as an `ExpertMLP` that REAP can score and structurally prune.

This is a stacked draft PR based on #2978 because the Kimi-K3 model definition is introduced there and is not yet available on `main`. After #2978 merges, this branch can be rebased and the base changed to `main`.

### Standardize Kimi-K3 routed experts

- Add `KimiLinearExperts`, a `LinearExperts2D` subclass containing the Kimi routed experts.
- Change `KimiBlockSparseMLP` to inherit from `ExpertMLPWithGate`.
- Preserve the Kimi-specific SiTU activation, standard SiLU gated activation, latent expert hidden size, and bias-free projections.
- Keep `KimiSparseMoeBlock` as the owner of routing, shared experts, and latent down/up projections.
- Replace its custom `moe_train` and `moe_infer` dispatch paths with `LinearExperts2D.forward(hidden_states, topk_ids, topk_weights)`.

### Preserve checkpoint compatibility

Register reversible Transformers weight renames for routed experts:

- `w1` -> `gate_proj`
- `w2` -> `down_proj`
- `w3` -> `up_proj`

The mappings allow the original Kimi-K3 checkpoint layout to load into the standardized runtime modules and restore the original names when saving.

### Support Kimi-K3 in REAP

- Recognize `num_experts_per_token` as a top-k config field.
- Recognize `num_expert_group` as an expert-group config field.
- Preserve `e_score_correction_bias` as an `nn.Parameter` when the router is resized. Existing routers that register it as a buffer retain buffer behavior.

### Tests and example

- Add module-level equivalence coverage against the previous Kimi expert dispatch.
- Verify the checkpoint mappings round-trip for `w1`, `w2`, and `w3`.
- Verify Kimi config discovery and correction-bias pruning.
- Load `inference-optimization/Kimi-K3-0.40B` through both the original remote definition and the LLM Compressor definition and compare logits.
- Add a 25% REAP example for the tiny Kimi-K3 checkpoint.

## Duplicate-work check

No open PR references #2979, and no open PR was found for Kimi-K3 REAP or `LinearExperts2D` integration. Draft PR #2978 provides the prerequisite Kimi-K3 model definition and quantization work; it does not implement REAP integration, checkpoint renaming to `LinearExperts2D`, or expert pruning.

## Validation

```text
PYTHONPATH=src /tmp/llmc-dev-venv/bin/python -m pytest tests/llmcompressor/modeling/test_kimi_k3.py -q
5 passed

PYTHONPATH=src /tmp/llmc-dev-venv/bin/python -m pytest tests/llmcompressor/modifiers/pruning/reap/test_base.py -q
24 passed

RUFF_CACHE_DIR=/tmp/issue2979/ruff-cache ruff check --ignore E501 <changed files>
All checks passed

git diff --check -- <changed files>
Passed
```

A four-sample GPU REAP smoke run completed end-to-end on `inference-optimization/Kimi-K3-0.40B`. REAP discovered all seven MoE layers and pruned each layer from 8 experts to 6 at 25% sparsity.

The smoke run used the matching `compressed-tensors` dependency commit `7074ea1` from `kylesayrs/july28`; #2978 currently depends on its `leave_decompressed` API, which is not yet on `compressed-tensors/main`.

## Model-output validation

The unpruned local definition matches the original remote definition within FP16 accumulation-order tolerance (`atol=5e-3`, `rtol=2e-2`). The observed maximum absolute difference was `0.00390625`, with 10 differing values out of 655,360 logits at a stricter `1e-3` tolerance.

No downstream accuracy benchmark was run for the pruned checkpoint; the end-to-end run validates algorithm execution and structural pruning only.

## AI assistance

AI assistance was used to analyze issue #2979, implement the changes, write tests and the example, run validation, and draft this PR description. The human submitter will review every changed line and is responsible for understanding and defending the change before marking the PR ready for review.

### CODE REVIEW

File: src/llmcompressor/modeling/kimi_k3/modeling_kimi_k3_linear.py [hunk 1]
```diff
@@ -274,42 +305,67 @@ def forward(self, hidden_states):
...
+            self,
+            [
+                KimiBlockSparseMLP(
+                    config,
+                    hidden_size=hidden_dim,
+                    intermediate_size=self.intermediate_size,
+                )
+                for _ in range(self.num_experts)
+            ],
+        )
```
[the remark this answers is not in the document]
Comment (KKothuri) [reply src/llmcompressor/modeling/kimi_k3/modeling_kimi_k3_linear.py:369]: Addressed in fef1f63f. KimiLinearExperts now initializes act_fn, alpha, and limit, matching the LinearExperts2D public interface; the unit test covers these attributes.

File: src/llmcompressor/modeling/kimi_k3/modeling_kimi_k3_linear.py [hunk 2]
```diff
@@ -274,42 +305,67 @@ def forward(self, hidden_states):
...
         else:
-            self.act_fn = ACT2FN[config.hidden_act]
+            act_fn = ACT2FN[config.hidden_act]
+
+        super().__init__(
+            hidden_dim=self.hidden_dim,
+            intermediate_size=self.ffn_dim,
+            mlp_bias=False,
+            _apply_gate=self._apply_gate,
+            dtype=config.dtype,
```
[the remark this answers is not in the document]
Comment (KKothuri) [reply src/llmcompressor/modeling/kimi_k3/modeling_kimi_k3_linear.py]: Addressed in fef1f63f. The expert now reads config.dtype safely and falls back to config.torch_dtype when unavailable.

**Review by kylesayrs — COMMENTED** [review]
Looks awesome so far!

**Comment by Ryfernandes** [comment]
Ditto to that! Love how simple the updates to REAP are
